Fix unordered releases bug

Pluies opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Fix incorrect latest version for github/gitlab upstreams

Extract the logic to get the latest version from a list of releases/tags and test it separately.

#### Which issue(s) this PR fixes:

Fixes #758

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Fix incorrect logic for latest version for github/gitlab upstreams when releases are unordered
```

cpanato commented:
Will need the cla from all authors

saschagrunert commented on upstream/upstream.go:
```diff
@@ -66,3 +70,35 @@ const (
+		}
+
+		if !expectedRange(version) {
+			log.Debugf("Skipping release not matching range constraints (%s): %s\n", constraints, tag)
```
There is no need for the newline on logrus statements:

```suggestion
			log.Debugf("Skipping release not matching range constraints (%s): %s", constraints, tag)
```

saschagrunert commented on upstream/upstream.go:
```diff
@@ -66,3 +70,35 @@ const (
+	candidateVersionString := "" // keep the version string separately as it may contain a leading `v`
+	for _, tag := range tags {
+		// Try to match semver and range
+		version, err := semver.Parse(strings.Trim(tag, "v"))
```
We can use `util.TagStringToSemver` for that: https://github.com/kubernetes-sigs/release-utils/blob/441a2c4cd8b97d693318f3c207752df8335db66b/util/common.go#L380

saschagrunert commented on upstream/upstream.go:
```diff
@@ -66,3 +70,35 @@ const (
+			continue
+		}
+
+		log.Debugf("Found potential release: %s\n", version.String())
```
There is no need for the newline on logrus statements:
```suggestion
		log.Debugf("Found potential release: %s", version.String())
```

saschagrunert commented on upstream/upstream.go:
```diff
@@ -66,3 +70,35 @@ const (
+
+		log.Debugf("Found potential release: %s\n", version.String())
+		if candidateVersionString == "" || version.GT(candidateVersion) {
+			log.Debugf("Release is the newest found so far: %s\n", version.String())
```
There is no need for the newline on logrus statements:
```suggestion
			log.Debugf("Release is the newest found so far: %s", version.String())
```

Pluies commented:
Closing in favour of #763

Pluies commented:
@saschagrunert thanks for the review 🙏 I've fixed things and removed the separate contribution but I'm not allowed to force-push to this branch, so I created #763 instead
